File: backend/app/schemas/simulation.py
# backend/app/schemas/simulation.py
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel

from app.schemas.scenario import Scenario, Direction
from app.schemas.fsm import FSMDefinition
from app.schemas.project import ElevatorConfig


class SimulationRequest(BaseModel):
    """
    Внутренний запрос симуляции: backend сам собирает его из ProjectConfig.
    """
    project_id: int
    config: ElevatorConfig
    fsm: FSMDefinition
    scenario: Scenario


class TimelineItem(BaseModel):
    time: int
    floor: int
    state_id: str
    doors_open: bool
    direction: Direction


class SimulationMetrics(BaseModel):
    avg_wait_time: float
    total_moves: int
    stops: int


class SimulationResult(BaseModel):
    timeline: List[TimelineItem]
    metrics: SimulationMetrics


class ProjectSimulationRequest(BaseModel):
    """
    То, что приходит в эндпоинт /projects/{id}/simulate с фронта.
    Все поля опциональны, можно просто отправить {}.
    """
    scenario: Optional[Scenario] = None
    config_override: Optional[ElevatorConfig] = None

File: frontend/src/api/simulation.ts
// src/api/simulation.ts
import { apiClient } from "./client";

export interface SimulationRequest {
  // Пока оставим как unknown — фронт будет подставлять структурированные данные позже
  project_id?: number;
  config: unknown;
  fsm: unknown;
  scenario: unknown;
}

export interface SimulationResult {
  timeline: Array<{
    time: number;
    floor: number;
    state_id: string;
    doors_open: boolean;
    direction: string;
  }>;
  metrics: {
    avg_wait_time: number;
    total_moves: number;
    stops: number;
  };
}

export interface ElevatorFrame {
  time: number;
  floor: number;
  state_id: string;
  doors_open: boolean;
  direction: string;
}


export const runSimulation = async (
  projectId: number
): Promise<SimulationResult> => {
  const { data } = await apiClient.post<SimulationResult>(
    `/projects/${projectId}/simulate`,
    {}
  );
  return data;
};
